### PR TITLE
command line option --json-file for internal-dpl-AOD-writer

### PR DESCRIPTION
**Pull request by pbuehler**

### CODE REVIEW

File: Framework/Core/include/Framework/DataOutputDirector.h [hunk 1]
```diff
@@ -17,6 +17,10 @@
 #include "Framework/DataSpecUtils.h"
 #include "Framework/InputSpec.h"
 
+#include "rapidjson/document.h"
+#include "rapidjson/filereadstream.h"
```
Comment (ktf) [inline Framework/Core/include/Framework/DataOutputDirector.h]: Can you forward declare this?

File: Framework/Core/include/Framework/DataOutputDirector.h [hunk 2]
```diff
@@ -17,6 +17,10 @@
...
+
+#include <cstdio>
```
Comment (ktf) [inline Framework/Core/include/Framework/DataOutputDirector.h]: Why is this needed?

File: Framework/Core/include/Framework/DataOutputDirector.h (hunk 1 above)
Comment (pbuehler) [reply Framework/Core/include/Framework/DataOutputDirector.h]: tried to clean this up

File: Framework/Core/include/Framework/DataOutputDirector.h (hunk 2 above)
Comment (pbuehler) [reply Framework/Core/include/Framework/DataOutputDirector.h]: is removed

File: Framework/Core/include/Framework/DataOutputDirector.h [hunk 3]
```diff
@@ -17,14 +17,13 @@
 #include "Framework/DataSpecUtils.h"
 #include "Framework/InputSpec.h"
 
-#include <regex>
+#include "rapidjson/document.h"
```
Comment (ktf) [inline Framework/Core/include/Framework/DataOutputDirector.h]: Do you really need to expose it in a public header?

File: Framework/Core/src/CommonDataProcessors.cxx [hunk 4]
```diff
@@ -207,20 +207,41 @@ DataProcessorSpec
...
+    std::string fnbase("AnalysisResults");
+    std::string filemode("RECREATE");
+    int ntfmerge = 1;
+
+    // values from json
+    auto fnjson = ic.options().get<std::string>("json-file");
+    if (!fnjson.empty()) {
+      auto [fnb, fmo, ntfm] = dod->readJson(fnjson);
+      if (!fnb.empty())
+        fnbase = fnb;
```
Comment (ktf) [inline Framework/Core/src/CommonDataProcessors.cxx]: Please use `{}` around the if, as per coding conventions.

File: Framework/Core/include/Framework/DataOutputDirector.h (hunk 3 above)
Comment (pbuehler) [reply Framework/Core/include/Framework/DataOutputDirector.h]: what exactly is the problem with this?

Comment (ktf) [reply Framework/Core/include/Framework/DataOutputDirector.h]: That we should try to avoid exposing unneeded dependencies, because it will bite us back. Can you use a `rapidjson/fwd.h` for example?

Comment (pbuehler) [reply Framework/Core/include/Framework/DataOutputDirector.h]: done

File: Framework/Core/src/CommonDataProcessors.cxx (hunk 4 above)
Comment (pbuehler) [reply Framework/Core/src/CommonDataProcessors.cxx]: Added brackets around single-line if and for statements - although I couldn't find the instruction in the coding guidelines. Anyway, it is saver and therefore better I agree.

**Comment by ktf** [comment]
The coding guidelines are at https://github.com/AliceO2Group/CodingGuidelines

**Comment by aphecetche** [comment]
And that specific rule is in the "Naming and formatting conventions" part
https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=Braces#Braces

**Comment by pbuehler** [comment]
> And that specific rule is in the "Naming and formatting conventions" part
> https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=Braces#Braces

thanks for the clarification!

**Comment by ktf** [comment]
@MichaelLettrich @davidrohr the test for the GPU sorting is still noisy. Did you have time to have a look at what is going on?

**Comment by ktf** [comment]
Merging this since it provides some extra feature. Notice however that your previous development seems to have broken some of the previous workflows, because you seem to create data loops in your aod-writer. In particular you seem to have multiple input specs with the same description, which is not allowed. Could you give a look, e.g. with the filters example?

**Comment by davidrohr** [comment]
@ktf @MichaelLettrich : Fix is here: alisw/alidist#2168
